### PR TITLE
Stateless schedulers

### PR DESCRIPTION
**Pull request by jbloxham**
Refactor of schedulers - the Composer Trainer can now accept stateless trainers of the form `(state: State) -> float`. Also, replaces all the pre-existing Composer schedulers with implementations in this format.

I still need to write an unholy amount of docstrings but I wanted to get eyes on this first.

### CODE REVIEW

File: composer/trainer/scale_schedule.py [hunk 1]
```diff
@@ -1,15 +1,50 @@
...
+        scheduler.gamma = scheduler.gamma**factor  # type: ignore  -- unknown attribute
+    else:
+        raise ValueError(f'Scale schedule being applied to unrecognized Scheduler {scheduler}. '
+                         'Please implement your scheduler as a function instead.')
+
+    return scheduler
+
+
+def _scale_composer_scheduler(scheduler: ComposerSchedulerFn, ssr: float):
+    return functools.partial(scheduler, ssr=ssr)  # type: ignore
```
Comment (jbloxham) [inline composer/trainer/scale_schedule.py:42]: This is another thing I'm not a huge fan of. It's absolutely necessary for `ssr` to be a scheduler fn param, but I wasn't sure whether I wanted that to be reflected within the `ComposerSchedulerFn` type. I went with leaving the `ssr` param out, which is why the `# type: ignore` is necessary here.

**Review by ravi-mosaicml — COMMENTED** [review]
Overall love it :100: 

Main suggestion would be to make the `ComposerScheduler` a protocol class rather than a function, mainly for typing reasons (I don't think it would impact at all how users would use it, since Protocol classes with `__call__` are equivalent to `Callable`).

Also, curious are we going to have a `ComposedScheduler` class? I guess a user can always write their own composed scheduler provided, but it might be nice to provide one.

File: composer/optim/scheduler.py [hunk 2]
```diff
@@ -1,226 +1,265 @@
...
+    elif ssr != 1.0:
+        time = Time(value=int(ssr * time.value), unit=time.unit)
+
+    if time.unit == TimeUnit.EPOCH:
+        time = convert_time(time=time, unit=TimeUnit.BATCH, steps_per_epoch=state.steps_per_epoch)
+
+    return time
+
+
+def step_scheduler(state: State, *, ssr: float = 1.0, step_size: Union[str, Time], gamma: float = 0.1) -> float:
```
[the remark this answers is not in the document]
Comment (hanlint) [reply composer/optim/scheduler.py:35]: Alternate opinion: classes seem overkill, stateless functions are preferred for simplicity. I've always found functools.partial has always been difficult to grok/read.

File: composer/core/state.py [hunk 3]
```diff
@@ -174,7 +175,11 @@ def __init__(
         if schedulers is None:
             self._schedulers = []
         else:
-            self._schedulers = list(ensure_tuple(schedulers))
+            from composer.core.scheduler import compile_scheduler
+            self._schedulers = [
+                scheduler if isinstance(scheduler, types.Scheduler) else compile_scheduler(
+                    scheduler, self.optimizers, self) for scheduler in ensure_tuple(schedulers)
+            ]
```
[the remark this answers is not in the document]
Comment (ajaysaini725) [reply composer/core/state.py:182]: Yeah +1 prob better to compile the schedulers in the trainer init

File: composer/optim/scheduler.py [hunk 4]
```diff
@@ -1,226 +1,265 @@
...
+
+
+def linear_with_warmup_scheduler(state: State,
+                                 *,
+                                 ssr: float = 1.0,
+                                 warmup_time: Union[str, Time],
+                                 start_factor: float = 1.0,
+                                 end_factor: float = 0.0,
+                                 total_time: Union[str, Time] = '1dur'):
+    # N.B. warmup time is intentionally *not* subject to scale schedule
```
Comment (ajaysaini725) [inline composer/optim/scheduler.py:173]: N.B. ? 

Should probably just include the detail about warmup in the doctoring for this function (+ all others that deal with warmup like this)

File: composer/optim/scheduler.py [hunk 5]
```diff
@@ -1,226 +1,265 @@
...
+        return linear_scheduler(state, start_factor=0.0, end_factor=1.0, total_time=warmup_time)
+
+    T_max = _convert_time(T_max, state, ssr=ssr)
+    current_time = state.timer.get(warmup_time.unit)
+    frac_of_total = ((current_time - warmup_time) / (T_max - warmup_time)).value
+
+    return _cosine_anneal(x=frac_of_total, min_y=min_factor)
+
+
+def get_scheduler(scheduler: ComposerSchedulerFn, ssr: float = 1.0, **kwargs) -> ComposerSchedulerFn:
```
Comment (ajaysaini725) [inline composer/optim/scheduler.py:205]: Maybe rename to `get_scheduler_with_ssr` to be more clear about what it's used for?

File: composer/yamls/models/bert-base.yaml [hunk 6]
```diff
@@ -42,20 +42,11 @@ optimizer:
...
-      warmup_iters: 0.06dur
-      warmup_factor: 0
-      interval: step
-  - linear_decay:
-      start_factor: 1.0
-      end_factor: 0.0
-      total_iters: 0.94dur
-      interval: step
-      verbose: false
+  - linear_decay_with_warmup:
```
Comment (ajaysaini725) [inline composer/yamls/models/bert-base.yaml:45]: What about the other parameters like `interval` and `verbose`?

File: composer/core/state.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbloxham) [reply composer/core/state.py:182]: Yup, works for me.

File: composer/trainer/scale_schedule.py (hunk 1 above)
Comment (jbloxham) [reply composer/trainer/scale_schedule.py:42]: I don't think having every scheduler arg be relative to the training duration is a good restriction. Warmup, especially, likes to run for a fixed absolute amount of steps, and also has weird behavior wrt SSR anyways.

File: composer/core/types.py [hunk 7]
```diff
@@ -52,8 +52,11 @@
 if TYPE_CHECKING:
     from typing import Protocol
 
+T = TypeVar('T')
+Many = Union[T, Tuple[T, ...], List[T]]
```
[the remark this answers is not in the document]
Comment (jbloxham) [reply composer/core/types.py:56]: Admittedly, I may be ditching this. I've been running into a lot of "ugh, Python" issues where e.g. a `Many[T]` cannot satisfy a `Many[Union[T, U]]`.

File: composer/core/scheduler.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+from torch.optim.lr_scheduler import LambdaLR
+
+from composer.core import State
+from composer.core.types import Optimizers, Scheduler
+from composer.utils.iter_helpers import ensure_tuple
+
+ComposerSchedulerFn = Callable[[State], float]
+
+
+def compile_scheduler(scheduler: ComposerSchedulerFn, optimizers: Optimizers, state: State) -> Scheduler:
```
[the remark this answers is not in the document]
Comment (jbloxham) [reply composer/core/scheduler.py:12]: Yeah fair point, I'm not really sure why I wrote it this way.

File: composer/optim/scheduler.py [hunk 9]
```diff
@@ -1,226 +1,265 @@
...
-                 last_epoch: int = -1,
-                 verbose: bool = False):
+    x = min(max(x, 0.0), 1.0)
+    return min_y + (max_y - min_y) * (1 + math.cos(x * math.pi)) / 2
+
+
+def cosine_annealing_scheduler(state: State,
+                               *,
+                               ssr: float = 1.0,
+                               T_max: Union[str, Time] = '1dur',
```
[the remark this answers is not in the document]
Comment (jbloxham) [reply composer/optim/scheduler.py:107]: yeah, im totally fine with that

File: composer/optim/scheduler.py (hunk 4 above)
Comment (jbloxham) [reply composer/optim/scheduler.py:173]: [Nota bene](https://en.wikipedia.org/wiki/Nota_bene) - I guess I can find a more normal way to say this. :joy:

File: composer/yamls/models/bert-base.yaml (hunk 6 above)
Comment (jbloxham) [reply composer/yamls/models/bert-base.yaml:45]: This PR admittedly removes support for `verbose: True`, but there's nowhere in our codebase that uses that.

`interval` is replaced by a new trainer param, `use_stepwise_schedulers`, which defaults to `True`. The only schedulers in these YAMLs which used `interval: epoch` were a few multistep schedulers, where it doesn't actually matter since their LRs only logically change on epoch boundaries anyways.

File: composer/optim/scheduler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbloxham) [reply composer/optim/scheduler.py:35]: I like the idea of defining the `ComposerScheduler` type as a `Protocol`, rather than a direct function type. However, I agree with Hanlin that it's a lot simpler for users if the implementations of our schedulers are pure functions.

File: composer/core/scheduler.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
+from typing import Callable
+
+from torch.optim.lr_scheduler import LambdaLR
+
+from composer.core import State
+from composer.core.types import Optimizers, Scheduler
+from composer.utils.iter_helpers import ensure_tuple
+
+ComposerSchedulerFn = Callable[[State], float]
```
[the remark this answers is not in the document]
Comment (jbloxham) [reply composer/core/scheduler.py:9]: It's tricky to have the trainer provide the SSR and state on every call if we stick to the current paradigm where the schedulers are passed to a `LambdaLR` object from Pytorch. The trainer isn't calling the schedulers directly. Partial shenanigans seem like the most obvious way to deal with this problem.

File: composer/optim/scheduler.py (hunk 5 above)
Comment (jbloxham) [reply composer/optim/scheduler.py]: Woops, this function doesn't need to be here at all actually.

File: composer/yamls/models/bert-base.yaml (hunk 6 above)
Comment (jbloxham) [reply composer/yamls/models/bert-base.yaml:45]: As for other params like, say, `total_iters` in this scheduler, they're no longer necessary and have been replaced by defaults.

File: tests/trainer/test_scheduler.py [hunk 11]
```diff
@@ -1,223 +1,111 @@
...
+    return State(
+        model=dummy_model,
+        train_dataloader=dummy_train_dataloader,
+        max_duration=MAX_DURATION,
+        steps_per_epoch=STEPS_PER_EPOCH,
+    )
+
+
+@pytest.mark.parametrize("scheduler,ssr,test_times,expected_lrs", [
+    pytest.param(functools.partial(step_scheduler, step_size='10ba'), 1.0, ['5ba', '15ba', '35ba'], [1.0, 0.1, 0.001]),
```
[the remark this answers is not in the document]
Comment (jbloxham) [reply tests/trainer/test_scheduler.py:32]: Yeah I knew I had to write a huge amount of tests, and this seemed like the best way to do so while keeping things not absurdly verbose. I definitely agree though that this is unreadable in its own way.

File: composer/optim/scheduler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/optim/scheduler.py:72]: Agreed; I think docstring examples can show that a scheduler can be implemented as a function, or as a class with `__call__`. Pyright should be happy with both styles :)

File: composer/core/scheduler.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/core/scheduler.py]: Sorry, don't quite follow. Wouldn't the compilation take care of binding the state / ssr and converting it to the lambda LR?

[the remark this answers is not in the document]
Comment (jbloxham) [reply composer/core/scheduler.py]: Ok, I think I just misunderstood you then - I do now have state and SSR both as arguments to `ComposerScheduler`.

File: composer/optim/scheduler.py [hunk 12]
```diff
@@ -1,226 +1,265 @@
...
-        last_epoch (int): the index of the last epoch. Can be used to restore the learning rate schedule state.
-            Default: ``-1``
-        verbose (bool): If ``True``, prints a message to stdout for each update. Default: ``False``.
+def linear_scheduler(state: State,
+                     *,
+                     ssr: float = 1.0,
+                     start_factor: float = 1.0 / 3,
+                     end_factor: float = 1.0,
+                     total_time: Union[str, Time] = '1dur') -> float:
+    total_time = _convert_time(total_time, state, ssr=ssr)
```
[the remark this answers is not in the document]
Comment (jbloxham) [reply composer/optim/scheduler.py:114]: I'm gonna opt not to do this here, for a few reasons:

1) It's a lot of extra code, when in theory the tests I wrote in `test_scheduler` should be sufficient to test correctness of these schedulers
2) In practice, the schedulers that we use the most (`cosine_annealing_with_warmup`, `linear_with_warmup`, and `polynomial`) don't have Pytorch equivalents
3) Equivalence isn't necessarily a design goal of these schedulers - for example, some params have been renamed, or had their behavior changed to better work with our time model

**Comment by jbloxham** [comment]
This is ready for final review - I'm going to start writing docstrings for this but I think it's probably best that docstrings don't block this from getting merged. The functionality in this PR could really do with a bit of time to bake.

File: composer/optim/scheduler.py [hunk 13]
```diff
@@ -1,213 +1,390 @@
...
-                         dataset_num_tokens: Optional[int] = None) -> None:
-    """Converts all fields in ``kwargs`` that were provided as timestrings (e.g. "32ep") into integers, representing
-    either epochs or batches, depending on the ``interval``.
-
-    Modifies ``kwargs`` in place.
+
+class ComposerSchedulerFn(Protocol):
+    """Specification for a "stateless" scheduler function.
+
+    A scheduler function should be a pure function that returns a learning rate, given the current
```
Comment (hanlint) [inline composer/optim/scheduler.py:32]: is it returning the learning rate, or a learning rate factor?

File: composer/optim/scheduler.py [hunk 14]
```diff
@@ -1,213 +1,390 @@
...
 @dataclass
 class PolynomialLRHparams(SchedulerHparams):
     """Hyperparameters for the :class:`PolynomialLR` scheduler."""
-    T_max: str = hp.required(doc='Maximum number of iterations.')
     power: float = hp.required(doc='Power of LR schedule.')
-    eta_min: float = hp.optional(default=0.0, doc='Minimum learning rate.')
-    verbose: bool = hp.optional(default=False, doc='Prints message to stdout.')
-    interval: str = hp.optional(default='step', doc=_interval_doc)
+    t_max: str = hp.optional(default='1dur', doc='Maximum number of iterations.')
+    min_factor: float = hp.optional(default=0.0, doc='Minimum learning rate.')
```
Comment (hanlint) [inline composer/optim/scheduler.py:364]: minimum learning rate factor?

File: tests/trainer/test_scheduler.py [hunk 15]
```diff
@@ -1,223 +1,112 @@
...
+def dummy_schedulers_state(dummy_model: Model, dummy_train_dataloader: DataLoader):
+    return State(
+        model=dummy_model,
+        train_dataloader=dummy_train_dataloader,
+        max_duration=MAX_DURATION,
+        steps_per_epoch=STEPS_PER_EPOCH,
+    )
+
+
+@pytest.mark.parametrize("scheduler,ssr,test_times,expected_lrs", [
```
Comment (hanlint) [inline tests/trainer/test_scheduler.py:31]: 🤯

Comment (hanlint) [reply tests/trainer/test_scheduler.py:31]: We can eventually clean this up with the `pytest_generate_tests` hook, and keep the data here in a dictionary, but not required for this PR. Or adopt the pytorch approach (https://github.com/pytorch/pytorch/blob/03662b32d55b33eb6a2d63bb8ccec4671e9d323f/test/test_optim.py#L1150), which i actually find a bit too verbose.

**Review by hanlint — COMMENTED** [review]
This is 🔥 

We should solicit some help in testing this. I'll chat with folks tomorrow about possible doing some runs with this to make sure the learning rate schedules match.

**Comment by growlix** [comment]
Wow, great work Jamie! I don't have much to contribute on top of Hanlin and Ajay. Happy to launch some test runs though!

**Review by hanlint — COMMENTED** [review]
minor nit

File: composer/trainer/trainer.py [hunk 16]
```diff
@@ -401,9 +388,26 @@ def __init__(
             evaluators=self.evaluators,
             optimizers=optimizers,
             steps_per_epoch=train_subset_num_batches,
-            schedulers=schedulers,
         )
 
+        if not schedulers:
+            schedulers = cosine_annealing_scheduler
```
[the remark this answers is not in the document]
Comment (jbloxham) [reply composer/trainer/trainer.py:394]: Yup, I'm fine with that.

File: composer/optim/scheduler.py [hunk 17]
```diff
@@ -1,213 +1,390 @@
...
+    elif ssr != 1.0:
+        time = Time(value=int(ssr * time.value), unit=time.unit)
+
+    if time.unit == TimeUnit.EPOCH:
+        time = convert_time(time=time, unit=TimeUnit.BATCH, steps_per_epoch=state.steps_per_epoch)
+
+    return time
+
+
+def compile_scheduler(scheduler: ComposerScheduler, state: State) -> Scheduler:
```
[the remark this answers is not in the document]
Comment (jbloxham) [reply composer/optim/scheduler.py:61]: I don't think this is right. `ComposerScheduler` is a union of `ComposerSchedulerFn` and `Scheduler`.

File: composer/trainer/scale_schedule.py [hunk 18]
```diff
@@ -1,15 +1,50 @@
...
 
 from torch.optim.lr_scheduler import CosineAnnealingLR, CosineAnnealingWarmRestarts, ExponentialLR, MultiStepLR, StepLR
 
 from composer.core.types import Scheduler
-from composer.optim.scheduler import ConstantLR
+from composer.optim.scheduler import ComposerSchedulerFn
 
 
-def scale_scheduler(scheduler: Scheduler, ssr: float, orig_max_epochs: Optional[int] = None):
+def _scale_pytorch_scheduler(scheduler: Scheduler, ssr: float, orig_max_epochs: Optional[int] = None):
```
[the remark this answers is not in the document]
Comment (jbloxham) [reply composer/trainer/scale_schedule.py:13]: Yeah I'm down to eventually phase out the old scale schedule code, but I think it's worth letting it bake for a little bit.

**Review by hanlint — APPROVED** [review]
The updated code looks good -- a few nits left but approving now as they are up to your discretion.
